src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -777,18 +780,17 @@ object ICustomAstVisitor.VisitFileRedirection(FileRedirectionAst fileRedirection
             return TypeInferenceContext.EmptyPSTypeNameArray;
         }
 
-        private IEnumerable<PSTypeName> InferTypesFrom(CommandAst commandAst)
+        private void InferTypesFrom(CommandAst commandAst, List<PSTypeName> inferredTypes)
         {
```

Comment:
Why the change from IEnumerable to List?

Reply:
I don't understand too why we add the allocations.

Reply:
Mostly debugability. It was a nightmare with all the nested state machines.

But since you have an allocation and virtual dispatch on every generated statemachine class that is generated when we use yield, my guess is that we actually have fewer allocations now.

And in general, I think the code got simpler. Several instances of loops that just yielded the loop object that just disappeared.

If you want to do the exercise, try debugging a tab completion before and after the change.

Reply:
Yes, debugging can be a headache here.

As for allocations my understanding is that it depends on how we access the object - if we use IEnumerable interface we allocate in both cases, if we cast to a collection we don't allocate I believe.

Reply:
But there is no underlying collection to cast to. All the methods here that yield only has the generated state machine classes.